cpp/include/ray/api/ray_register.h

```diff
@@ -0,0 +1,201 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <ray/api/serializer.h>
+#include "absl/utility/utility.h"
+
+#include <boost/callable_traits.hpp>
+#include <functional>
+#include <string>
+#include <unordered_map>
+
+namespace ray {
+namespace api {
+
+enum ErrorCode {
+  OK = 0,
+  FAIL = 1,
+};
+
+struct VoidResponse {
+  int error_code;
+  std::string error_msg;
+
+  MSGPACK_DEFINE(error_code, error_msg);
+};
+
+template <typename T>
+struct Response {
+  int error_code;
+  std::string error_msg;
+  T data;
+
+  MSGPACK_DEFINE(error_code, error_msg, data);
+};
+
+template <class, class>
+struct AddType;
+
+template <class First, class... Second>
+struct AddType<First, std::tuple<Second...>> {
+  using type = std::tuple<First, Second...>;
+};
+
+template <class First, class Second>
+using AddType_t = typename AddType<First, Second>::type;
+
+template <typename OUT, typename IN>
+inline static OUT AddressOf(IN in) {
+  union {
+    IN in;
+    OUT out;
+  } u = {in};
+
+  return u.out;
+};
+
+class Router {
+ public:
+  static Router &Instance() {
+    static Router instance;
+    return instance;
+  }
+
+  template <typename Function>
+  bool RegisterHandler(std::string const &name, const Function &f) {
+    auto pair = func_ptr_to_key_map_.emplace(AddressOf<uintptr_t>(f), name);
+    if (!pair.second) {
+      return false;
+    }
+
+    RegisterNonmemberFunc(name, f);
+    return true;
+  }
+
+  std::string Route(const char *data, std::size_t size) {
+    std::string result;
+    do {
+      try {
+        auto p = Serializer::Deserialize<std::tuple<std::string>>(data, size);
+        auto &func_name = std::get<0>(p);
+        auto it = map_invokers_.find(func_name);
+        if (it == map_invokers_.end()) {
+          result = PackArgsStr(ErrorCode::FAIL, "unknown function: " + func_name);
+          break;
+        }
+
+        it->second(data, size, result);
+        if (result.size() >= MAX_BUF_LEN) {
+          result = PackArgsStr(
+              ErrorCode::FAIL,
+              "the response result is out of range: more than 10M " + func_name);
+        }
```

Comment:
Is this part of the final code? I don't think it's reasonable to limit the result size.

Reply:
Not the final code, we should consider some exception, if the user return a very large object maybe exhaust all the memory, i think we should limit the size, the max size can be more larger, whatever we need a capacity to keep safe call.

Reply:
If the object is large enough, we use plasma store to store the raw data and pass an object reference to the remote function. Same for the return value. So this max length check is unnessesary. We don't have this in Python or Java.